### ISSUE TITLE
SC2034 doesn't consider namerefs

### ISSUE THREAD

**Issue by zx2c4** [koalaman/shellcheck#817]
Shellcheck doesn't account for namerefs properly.

#### For bugs
- SC2034
- 4.5 and online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
#### Here's a snippet or screenshot that shows the problem:

```bash
#!/bin/bash
read_bool() {
        local -n out="$1"
        case "$2" in
        true) out=1 ;;
        false) out=0 ;;
        *) die "\`$2' is neither true nor false"
        esac
}
```

#### Here's what shellcheck currently says:

```
        false) out=0 ;;
               ^-- SC2034: out appears unused. Verify it or export it.
```

#### Here's what I wanted or expected to see:

Nothing.

**Comment by zx2c4** [comment]
Hey @koalaman - might I poke you about this?

Namerefs are an important feature of bash.

**Comment by koalaman** [comment]
No updates, but you're right. Namerefs should be considered auto-used just like exported variables.

**Comment by pabs3** [comment]
Duplicates of this issue: #1544 #1543

**Comment by hseg** [comment]
More apparent duplicates: #2919

**Comment by oklopfer** [comment]
7.5 year old issue now. Constant annoyance in our workflow. SC2034 is very helpful when it works correctly in identifying stray vars that need to be cleaned up, but cannot recognize any kind of nameref usage; not with a `declare/local -n`, not when adding entries to that ref, and not when using the ref in another function:
```
myfunc() {
    local -n out="${1}" str="${2}"
    str="${out[*]}"
}
array=(1 2 3)
declare -n ref="array"
ref+=(4)
myfunc ref string
echo "${string}"
```

![IMG_4291](https://github.com/user-attachments/assets/73e680f6-87e6-41ad-a428-bf264098d6b3)

**Comment by marc-hb** [comment]
https://www.shellcheck.net/wiki/SC2034 says

> This is an intentional design decision and not a bug. If you have variables that will not have direct references, consider using an associative array in bash, or just [ignore](https://www.shellcheck.net/wiki/ignore) the warning.

> Tracking indirect references is a common problem for compilers and static analysis tools, and it is known to be unsolvable in the most general case.

On the other hand, all the warning examples are about the "inside" name, not the outside name. They're also all reading, never writing. So, does this design choice extend to the example in the description above? I don't know.

**Comment by marc-hb** [comment]
A reminder that you do not have to disable the warning on the entire file. For instance this is shellcheck clean:

```shell
#!/bin/bash
read_bool() {
        local -n out="$1"
        case "$2" in
        true)
            out=1 ;;
        false)
            # shellcheck disable=SC2034
            out=0 ;;
        *) die "\`$2' is neither true nor false"
        esac
}

```
